status/pricing.py

```diff
@@ -583,109 +255,340 @@ def _validate_discontinued(self, components, products):
 
-        current_products = self.fetch_products(None)
+        self._validate_conserved(self.draft_components, self.published_components, 'components')
 
-        if current_products:
-            self._validate_conserved(products, current_products, 'products')
+        self._validate_unique(self.draft_products, 'products')
+        self._validate_not_null(self.draft_products, 'products')
+
+        self._validate_conserved(self.draft_products, self.published_products, 'products')
 
         # Verify no discontinued components are used for enabled products
-        self._validate_discontinued(components, products)
+        self._validate_discontinued(self.draft_components, self.draft_products)
+
+
+class PricingValidateDraftDataHandler(PricingBaseHandler):
+    """Loaded through /api/v1/pricing_validate_unsaved_draft """
+
+    def post(self):
+        draft_content = tornado.escape.json_decode(self.request.body)
+        published_doc = self.fetch_latest_published()
+
+        validator = PricingValidator(draft_content, published_doc)
+        validator.validate()
+        validator.track_all_changes()
+
+        response = dict()
+        response['validation_msgs'] = validator.validation_msgs
+        response['changes'] = validator.changes
+
+        self.set_header("Content-type", "application/json")
+        self.write(json.dumps(response))
+
+
+class PricingReassignLockDataHandler(PricingBaseHandler):
+    """Available at /api/v1/pricing_reassign_lock"""
+
+    def post(self):
+        doc = self.fetch_latest_doc()
+        draft = doc['Draft']
+        if not draft:
+            self.set_header("Content-type", "application/json")
+            self.set_status(400)
+            return self.write("Error: Attempting to update a non-draft cost calculator.")
+
+        user_email = self.get_current_user().email
+
+        lock_info = {'Locked': True,
+                     'Locked by': user_email}
+
+        doc['Lock Info'] = lock_info
+        doc = self._update_last_modified(doc)
+
+        cc_db = self.application.cost_calculator_db
+        cc_db.save(doc)
+        self.set_header("Content-type", "application/json")
+        self.write({'message': 'Lock info updated'})
+
+
+class PricingDraftDataHandler(PricingBaseHandler):
+    """Loaded through /api/v1/draft_cost_calculator """
 
     def get(self):
-        components = self.fetch_components()
-        products = self.fetch_products(None)
-        try:
-            self.validate(components, products)
-        except PricingValidationError:
-            raise tornado.web.HTTPError(
-                400, reason='\n'.join(self.validation_msgs))
-        except Exception:
-            raise # Maybe this is the default?
+        """ Should return the specified version of the cost calculator."""
+        doc = self.fetch_latest_doc()
+        draft = doc['Draft']
+
+        response = dict()
+        if draft:
+            response['cost_calculator'] = doc
+        if not draft:
```

Comment:
Is this check needed? There could be a default `response['cost_calculator'] = None` before `if draft:`

Reply:
Fixed!